tla-io/src/main/scala/at/forsyte/apalache/io/json/JsonDecoder.scala

```diff
@@ -5,10 +5,13 @@ import at.forsyte.apalache.tla.lir.{TlaDecl, TlaEx, TlaModule}
 /**
  * A JsonDecoder defines a conversion from a json (as represented by T) to a TLA+ expression/declaration/module
  *
+ * @tparam R
+ *   The class of the value which the JsonRepresentation uses to reprsent JSON
+ *
  * @tparam T
  *   Any class extending JsonRepresentation
  */
-trait JsonDecoder[T <: JsonRepresentation] {
+trait JsonDecoder[R, T <: JsonRepresentation[R]] {
```

Comment:
If you're going to parameterize `JsonRepresentation`, you could just have ` JsonDecoder[R]`, and then have the methods be e.g. `asTlaModule(moduleJson: JsonRepresentation[R]): TlaModule`. Unless you're thinking of having different subclasses of `JsonRepresentation[R]` for the same `R` (which I find unlikely).

Reply:
I fixed this a different way, using an abstract type member, but it much cleaner and less invasive than what I'd first proposed, so thanks for pushing back.